### PR TITLE
[3.x] Convert lustre.rb into a resource

### PR DESCRIPTION
**Pull request by eantonin**
### Description of changes
* Converted lustre.rb into a resource, splitting by OS
* I just moved the old lustre file, but still GH marked it as deleted
* I changed the RH8 Kernel release because if was falling into the Fsx issue branch

### Tests
* Docker tests 
* EC2 tests (RH8 was tested manually due to a known issue of FSx in a specific kernel version)

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cookbooks/aws-parallelcluster-install/attributes/default.rb [hunk 1]
```diff
@@ -4,3 +4,41 @@
 default['cluster']['sysctl']['ipv4']['gc_thresh1'] = 0
 default['cluster']['sysctl']['ipv4']['gc_thresh2'] = 15_360
 default['cluster']['sysctl']['ipv4']['gc_thresh3'] = 16_384
+
+# Lustre defaults (for CentOS >=7.7 and Ubuntu)
+default['cluster']['lustre']['public_key'] = value_for_platform(
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/attributes/default.rb]: `value_for_platform` is a hidden if and therefore it should be moved to the resource, here and below.

File: cookbooks/aws-parallelcluster-install/libraries/helpers.rb [hunk 2]
```diff
@@ -12,6 +12,37 @@
...
+#
+# Retrieve RHEL OS minor version from running kernel version
+# The OS minor version is retrieved from the patch version of the running kernel
+# following the mapping reported here https://access.redhat.com/articles/3078#RHEL7
+# Method works for CentOS7 minor version >=7
+#
+def find_centos_minor_version
+  os_minor_version = ''
+
+  if node['platform'] == 'centos'
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/libraries/helpers.rb]: OS-related `if`. Should be moved to a resource.

File: cookbooks/aws-parallelcluster-install/libraries/helpers.rb [hunk 3]
```diff
@@ -12,6 +12,37 @@
 # This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied.
 # See the License for the specific language governing permissions and limitations under the License.
 
+def kernel_release
+  ENV['KERNEL_RELEASE'] || default['cluster']['kernel_release']
+end
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/libraries/helpers.rb]: This looks like something belonging to `common`, isn't it?

File: cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+# Default resource implementation
+provides :lustre, platform: 'centos' do |node|
+  node['platform_version'].to_i == 7
+end
+
+use 'partial/_install_lustre_centos_redhat'
+use 'partial/_install_lustre_old_centos'
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb]: Why is this a partial? As far as I understand, this code belongs to centos...

File: cookbooks/aws-parallelcluster-install/resources/lustre/lustre_ubuntu2004.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file.
+# This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+
+provides :lustre, platform: 'ubuntu', platform_version: '20.04'
+unified_mode true
+
+use 'partial/_install_lustre_debian'
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/lustre_ubuntu2004.rb:19]: Nitpick. You can simplify the code by defining `setup` action in the partial and just including it here.

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_centos_redhat.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    retries 3
+    retry_delay 5
+  end
+
+  package %w(kmod-lustre-client lustre-client) do
+    retries 3
+    retry_delay 5
+  end
+
+  kernel_module 'lnet' unless virtualized?
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_centos_redhat.rb]: There was no check for `virtualized` in the old code. Why did you need to add it?

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_debian.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    retries 3
+    retry_delay 5
+  end unless virtualized?
+
+  package "lustre-client-modules-aws" do
+    retries 3
+    retry_delay 5
+  end
+
+  kernel_module 'lnet' unless virtualized?
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_debian.rb]: Same as above.

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_old_centos.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  lustre_kmod_rpm = "#{node['cluster']['sources_dir']}/kmod-lustre-client-#{node['cluster']['lustre']['version']}.x86_64.rpm"
+  lustre_client_rpm = "#{node['cluster']['sources_dir']}/lustre-client-#{node['cluster']['lustre']['version']}.x86_64.rpm"
+
+  # Get Lustre Kernel Module RPM
+  remote_file lustre_kmod_rpm do
+    source node['cluster']['lustre']['kmod_url']
+    mode '0644'
+    retries 3
+    retry_delay 5
+    not_if { ::File.exist?(lustre_kmod_rpm) }
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_old_centos.rb]: Nitpick: you can use `:create_if_missing` action to this end.

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_old_centos.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  # Get Lustre Client RPM
+  remote_file lustre_client_rpm do
+    source node['cluster']['lustre']['client_url']
+    mode '0644'
+    retries 3
+    retry_delay 5
+    not_if { ::File.exist?(lustre_client_rpm) }
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_old_centos.rb]: Nitpick. Same as above.

File: test/common/libraries/os_properties.rb [hunk 10]
```diff
@@ -27,6 +27,10 @@ def centos?
     inspec.os.name == 'centos'
   end
 
+  def centos_min_version?(version)
+    centos? && inspec.os.release.to_f >= version
+  end
```
Comment (awslitvit) [inline test/common/libraries/os_properties.rb]: I am conflicted about this. We should know which version we run the test on, right?

File: test/resources/controls/aws_parallelcluster_install/lustre_spec.rb [hunk 11]
```diff
@@ -0,0 +1,58 @@
+control 'lustre_client_installed' do
+  title "Verify that lustre client is installed"
+
+  kernel_release = command('uname -r').stdout.strip
```
Comment (awslitvit) [inline test/resources/controls/aws_parallelcluster_install/lustre_spec.rb]: Nitpick: could you define the variable close to where it is used?

File: test/common/libraries/os_properties.rb (hunk 10 above)
Comment (awslitvit) [reply test/common/libraries/os_properties.rb]: Also, I am not sure we need a library function. It looks very specific.

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_debian.rb [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    retries 3
+    retry_delay 5
+  end
+
+  apt_update
+
+  package "lustre-client-modules-#{node['cluster']['kernel_release']}" do
+    retries 3
+    retry_delay 5
+  end unless virtualized?
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_debian.rb]: Same as above.

**Review by awslitvit — CHANGES_REQUESTED** [review]

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_old_centos.rb (hunk 8 above)
Comment (eantonin) [reply cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_old_centos.rb]: Thanks, will fix

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_centos_redhat.rb (hunk 6 above)
Comment (eantonin) [reply cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_centos_redhat.rb]: I do not know how system tests are working now, but if I remove the virtualized the docker crashes

File: test/common/libraries/os_properties.rb (hunk 10 above)
Comment (eantonin) [reply test/common/libraries/os_properties.rb]: > We should know which version we run the test on, right?

We do, but we could change it and the test code must work (at least for the version in which we have the information on what to test)

> Also, I am not sure we need a library function. It looks very specific.

I can agree on that, moving it to the test code

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_debian.rb (hunk 12 above)
Comment (eantonin) [reply cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_debian.rb]: This is slightly different from the above case, as docker has a `KERNEL_RELEASE` value which is `3.10.0-1160.42.2.el7.x86_64` which does not map to an actual package. I could have added some if to force a specific version when on docker, but seemed overkill to me

And i don't want to change `KERNEL_RELEASE` in ubuntu docker as it could impact other tests

Doc https://docs.aws.amazon.com/fsx/latest/LustreGuide/install-lustre-client.html

File: cookbooks/aws-parallelcluster-install/resources/lustre/lustre_redhat8.rb [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+end
+unified_mode true
+
+use 'partial/_install_lustre_centos_redhat'
+
+default_action :setup
+
+action :setup do
+  if node['platform_version'].to_f < 8.2
+    Chef::Log.warn("Unsupported version of Red Hat, #{node['platform_version']}, supported versions are >= 8.2")
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/lustre_redhat8.rb]: Nitpick:
1. RHEL instead of Red Hat
2. I would mention that FSx for Lustre is not supported for this version, not the version itself.

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_centos_redhat.rb (hunk 6 above)
Comment (awslitvit) [reply cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_centos_redhat.rb]: We need to investigate this part because with this `unless`we remove something which was tested by existing system tests.

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_debian.rb (hunk 12 above)
Comment (awslitvit) [reply cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_debian.rb]: Could we use `KERNEL_RELEASE` from ubuntu AMIs?

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_centos_redhat.rb (hunk 6 above)
Comment (eantonin) [reply cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_centos_redhat.rb]: I agree on that, but I'd not do it now

File: cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+      source lustre_kmod_rpm
+    end
+
+    # Install lustre mount drivers
+    package 'lustre_client' do
+      source lustre_client_rpm
+    end
+  elsif node['platform_version'].to_f >= 7.7
+    execute 'yum-config-manager_skip_if_unavail' do
+      command "yum-config-manager --setopt=\*.skip_if_unavailable=1 --save"
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb]: Not sure if Francesco made this code reusable or not. If he did, I would advise to use `package_repo` resource.

Comment (eantonin) [reply cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb]: Yeah, I synced with Francesco here, and he suggested me to execute the command directly.
Reusing the code would have meant to add another "hidden" if

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_centos_redhat.rb (hunk 6 above)
Comment (eantonin) [reply cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_centos_redhat.rb]: Using mocks and installing packages I was able to remove the majority of virtualized statements,  I needed to leave a `redhat_ubi` one as I was not able to mock `dracut` command and it was trying to access kernel module functions which cannot work on docker

File: cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_debian.rb (hunk 12 above)
Comment (eantonin) [reply cookbooks/aws-parallelcluster-install/resources/lustre/partial/_install_lustre_debian.rb]: This should be fixed

File: cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+kmod_url_hash = {
+  '7.6' => "https://downloads.whamcloud.com/public/lustre/lustre-2.10.6/el7/client/RPMS/x86_64/kmod-lustre-client-2.10.6-1.el7.x86_64.rpm",
+  '7.5' => "https://downloads.whamcloud.com/public/lustre/lustre-2.10.5/el7.5.1804/client/RPMS/x86_64/kmod-lustre-client-2.10.5-1.el7.x86_64.rpm",
+}
+
+default_action :setup
+
+action :setup do
+  version = node['platform_version'].to_f
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb]: Just out of curiosity: why `to_f`?

File: cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+      source lustre_client_rpm
+    end
+
+    kernel_module 'lnet'
+
+  elsif node['platform_version'].to_f >= 7.7
+    execute 'yum-config-manager_skip_if_unavail' do
+      command "yum-config-manager --setopt=\*.skip_if_unavailable=1 --save"
+    end
+    action_install_lustre
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb:76]: Just out of curiosity again: did you check packages for `5.5` and `5.6` are still not available in the repository?

File: cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    # Install lustre mount drivers
+    package 'lustre_client' do
+      source lustre_client_rpm
+    end
+
+    kernel_module 'lnet'
+
+  elsif node['platform_version'].to_f >= 7.7
+    execute 'yum-config-manager_skip_if_unavail' do
+      command "yum-config-manager --setopt=\*.skip_if_unavailable=1 --save"
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb]: I wonder why you set this option before adding a new repository. Shouldn't it be done in `action_install_lustre` after adding the repo?

File: cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb (hunk 15 above)
Comment (eantonin) [reply cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb]: not needed for line 41, but it's needed for the second `if` using `>=7.7` (which was, mistakenly, not using the variable) will fix

File: cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb (hunk 16 above)
Comment (eantonin) [reply cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb:76]: I guess you mean `7.5` and `7.6`
I haven't done an explicit check but also in the public documentation https://docs.aws.amazon.com/fsx/latest/LustreGuide/install-lustre-client.html we can see a different path for downloading the needed packages in those versions of Centos

Comment (awslitvit) [reply cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb:76]: Sure, I meant `7.x`.
Thank you for the clarification.

File: cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb (hunk 17 above)
Comment (eantonin) [reply cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb]: You are totally right! Thanks, will fix

File: cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+    # Install lustre mount drivers
+    package 'lustre_client' do
+      source lustre_client_rpm
+    end
+
+    kernel_module 'lnet'
+
+  elsif version.to_f >= 7.7
+    action_install_lustre
+  elsif Chef::Log.warn("Unsupported version of Centos, #{node['platform_version']}, supported versions are >= 7.5")
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb:77]: Nitpick: could we use [log](https://docs.chef.io/resources/log/) resource?
It would be easier to test.

File: kitchen.resources.yml [hunk 19]
```diff
@@ -53,6 +53,19 @@ suites:
...
+      - recipe[aws-parallelcluster::add_dependencies]
+      - recipe[aws-parallelcluster-install::test_resource]
+    verifier:
+      controls:
+        - lustre_client_installed
+        - lnet_kernel_module_enabled
+    attributes:
+      dependencies:
+        - recipe:aws-parallelcluster-test::docker_mock
+        - resource:package_repos
```
Comment (awslitvit) [inline kitchen.resources.yml:67]: Is it still required?

File: test/resources/controls/aws_parallelcluster_install/lustre_spec.rb [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+      it { should be_installed }
+    end
+
+    if (os_properties.centos? && inspec.os.release.to_f >= 7.7) || os_properties.redhat?
+      describe yum.repo('aws-fsx') do
+        it { should exist }
+        it { should be_enabled }
+        its('baseurl') { should include 'fsx-lustre-client-repo.s3.amazonaws.com' }
+      end
+    end
```
Comment (awslitvit) [inline test/resources/controls/aws_parallelcluster_install/lustre_spec.rb:19]: Nitpick. Not sure we need this check. We need packages to be installed, not the repo to be configured. Why should it matter?

File: test/resources/controls/aws_parallelcluster_install/lustre_spec.rb [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  if os_properties.debian_family?
+    describe apt('https://fsx-lustre-client-repo.s3.amazonaws.com/ubuntu') do
+      it { should exist }
+      it { should be_enabled }
+    end
```
Comment (awslitvit) [inline test/resources/controls/aws_parallelcluster_install/lustre_spec.rb:26]: Same as above.

**Review by awslitvit — APPROVED** [review]

File: kitchen.resources.yml (hunk 19 above)
Comment (eantonin) [reply kitchen.resources.yml:67]: yes, it is for the installation of `yum_utils` needed by `yum-config-manager --setopt=\*.skip_if_unavailable=1 --save`

File: test/resources/controls/aws_parallelcluster_install/lustre_spec.rb (hunk 20 above)
Comment (eantonin) [reply test/resources/controls/aws_parallelcluster_install/lustre_spec.rb:19]: A thing that came to my mind is to ensure we download the packages from the right "authorative" repo that is specified in the documentation

https://docs.aws.amazon.com/fsx/latest/LustreGuide/install-lustre-client.html

anyway if we don't want to do this check we may remove the tests yes, won't do it in this PR

File: cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb (hunk 18 above)
Comment (eantonin) [reply cookbooks/aws-parallelcluster-install/resources/lustre/lustre_centos7.rb:77]: Sure, will do that in the next PR introducing ChefSpec for this resource
